pkg/yang/entry.go

```diff
@@ -925,6 +925,11 @@ func (e *Entry) Augment(addErrors bool) (processed, skipped int) {
 		// are merged into another entry.
 		processed++
 		ae.merge(nil, a.Namespace(), a)
+		if ae.Annotation == nil {
+			ae.Annotation = map[string]interface{}{}
+			ae.Annotation["Augmentation"] = []*Entry{}
+		}
+		ae.Annotation["Augmentation"] = append(ae.Annotation["Augmentation"].([]*Entry), a)
```

Comment:
For container and leaf `*Entry` we can get their when statements's XPath expression as follows:
`whenXPath, ok := *Entry.Node.(*Container/*Leaf).When.Statement().Arg()`
This is maybe a little verbose, but otherwise very workable.

 So I don't think a `type when struct` is necessary, because if we have the `*Entry`, then the `*Entry` == the `logicalProvenance` field, and the XPath `condition` can be extracted as demonstrated above. 

The problem with augment and uses is that the `*Entry` tree doesn't contain augment or uses `*Entry`, only the `*Entry` of the data nodes and/or groupings that are referred to by the augment and uses. Without access to augment and uses `*Entry` we also lose all information about them, including their when statements. So if we store augment and uses `*Entry` in the `*Entry` tree, then we have access to all the information the augment and uses `*Entry` have, not just the when statements.

What do you think?

Reply:
`*Entry.Node.(*Container).When.Statement().Arg()` is a bit wordy -- but we can probably help this by just having a helper that returns this on a particular `*Entry` receiver. So this sounds OK to me.

I have a concern with just copying the `*Entry` here though -- at this point, we've now included all the hierarchy below the `augment` `Entry` again in the tree. For example, if we have these modules:

```yang
module test {
  prefix "t";
  namespace "urn:t";

  container base { }
}
```

```yang
module test2 {
  prefix "t2";
  namespace "urn:t2";

  import test { prefix t; }
  
  augment /t:base {
    container a {
      container b {
        container c {
          leaf d { type string; }
        }
      }
    }
  }
}
```

Then the entry at `/t:base`'s `Augmented` will contain the `augment t:base` entry, and its `/t:base`'s `Dir` will also contain the same data. This will be for the entire tree.

It seems like it might be more efficient to, rather than store the whole entry, consider whether we can store a subset of it. (Note this information is something that we need to store in the JSON serialisation too - so there's a cost for tools that are using this to persist the processed tree - like ygot - so this is why I am concerned with the size.)

There are two options that spring to mind:

1) Store only the direct children of a particular Entry in the `Dir` of the entry which is included in `Augmented` or `Uses`.
2) Rather than store the `Entry` in `Augmented` -- add the `when` statements directly to entries that are included due to the `uses` or `augment`.

The latter approach would mean that:

```
container a {
   uses foo { when "something" }
}

grouping foo {
   leaf a { type string ; }
   container b { 
    leaf c { type string; }
   }
}
```

We would end up appending the `when` statement to the entry for `a` and `b` -- this results in duplication of the `when` contents -- and potentially processing by other tools.

In the former, we'd include only `foo` as an `Entry` in the `Uses` slice, but `c` would not be included. More efficient for processing - and avoids duplication of the entire tree when we serialise. We could actually remove anything in the `Entry` other than the children's name in this case.

Reply:
I have added a helper function `func (*Entry) GetWhenXPath() (string, bool)` as a shortcut for `*Entry.Node.(*TypeThatHaveWhenStatements).When.Statement().Arg()`

I went with option 1: Store only the direct children of a particular Entry in the `Dir` of the entry which is included in `Augmented` or `Uses`. I added a helper function `shallowDup` for this purpose.

Thanks for the timely and detailed responses. Let me know what you think of the latest version.